crates/bitcoin-da/src/helpers/parsers.rs

```diff
@@ -71,111 +84,109 @@ fn get_script(tx: &Transaction) -> Result<&Script, ParserError> {
-                        || (inside_envelope_index == 6 && bytes.as_bytes() != RANDOM_TAG)
-                        || (inside_envelope_index == 8 && bytes.as_bytes() != BODY_TAG)
-                    {
-                        return Err(ParserError::EnvelopeHasIncorrectFormat);
-                    } else if inside_envelope_index == 1
-                        && bytes.as_bytes() != rollup_name.as_bytes()
-                    {
-                        return Err(ParserError::InvalidRollupName);
-                    } else if inside_envelope_index == 3 {
-                        signature.extend(bytes.as_bytes());
-                    } else if inside_envelope_index == 5 {
-                        public_key.extend(bytes.as_bytes());
-                    } else if inside_envelope_index >= 9 {
-                        body.extend(bytes.as_bytes());
-                    }
-
-                    inside_envelope_index += 1;
-                } else if bytes.is_empty() {
-                    last_op = Some(OP_FALSE); // rust bitcoin pushes [] instead of op_false
-                }
-            }
-            Instruction::Op(another_op) => {
-                // don't allow anything except data pushes inside envelope
-                if inside_envelope {
-                    return Err(ParserError::EnvelopeHasNonPushOp);
-                }
-
-                last_op = Some(another_op);
-            }
+    // Parse header
+    let header_slice = read_push_bytes(instructions)?;
+    let Some(header) = TransactionHeader::from_bytes(header_slice.as_bytes()) else {
+        return Err(ParserError::InvalidHeaderLength);
+    };
+
+    // Check rollup name
+    if header.rollup_name != rollup_name.as_bytes() {
+        return Err(ParserError::InvalidRollupName);
+    }
+
+    // Parse transaction body according to type
+    match header.typ {
+        TransactionType::Inscribed => parse_type_0_body(instructions),
+        TransactionType::Unknown(_) => Err(ParserError::InvalidHeaderType),
+    }
+}
+
+fn read_instr<'a>(
+    instructions: &mut dyn Iterator<Item = Result<Instruction<'a>, ParserError>>,
+) -> Result<Instruction<'a>, ParserError> {
+    let instr = instructions
+        .next()
+        .unwrap_or(Err(ParserError::UnexpectedEndOfScript))?;
+    Ok(instr)
+}
+
+fn read_push_bytes<'a>(
+    instructions: &mut dyn Iterator<Item = Result<Instruction<'a>, ParserError>>,
+) -> Result<&'a StructPushBytes, ParserError> {
+    let instr = read_instr(instructions)?;
+    match instr {
+        PushBytes(push_bytes) => Ok(push_bytes),
+        _ => Err(ParserError::UnexpectedOpcode),
+    }
+}
+
+fn read_opcode(
+    instructions: &mut dyn Iterator<Item = Result<Instruction<'_>, ParserError>>,
+) -> Result<Opcode, ParserError> {
+    let instr = read_instr(instructions)?;
+    let Op(op) = instr else {
+        return Err(ParserError::UnexpectedOpcode);
+    };
+    Ok(op)
+}
```

Comment:
do we benefit from these read_* functions being inline?

Reply:
This is for readability only. But I believe rustc should inline them.